ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,496 @@
+  All common encoding converters
+* `Temporalio::Converters::FailureConverter` - Default failure converter, can be overridden
+* `Temporalio::Converters::PayloadCodec` - Unimplemented base codec
+  * ❓How can we make this async capable? We need to be able to multiplex several calls to this concurrently, but we may
+    not be in an async-capable environment and/or don't know how implementations may want to multiplex, so we can't just
+    have a naive payloads-to-payloads implementation. We probably need to allow the payload to accept multiple "payload
+    sets" and return "payload sets", since each set may contain multiple payloads, but sets need to be independent.
+* `Temporalio::Converters::RawValue` - Can be used as way to move around unconverted payloads.
+
+#### Temporalio::Converter::PayloadConverter::JSONPlain
+
+* Ruby has a `JSON` module in the standard library, but by default it's only primitives, arrays, hashes, etc and not
+  classes.
+  * There is a concept of "additions", but they set a `json_class` field on the object which makes the JSON not
+    cross-language compatible.
+  * There are not any good JSON-to-object conversion approaches in Ruby beyond this.
+  * Therefore, very unfortunately, it is unreasonable at this time for any area where conversion from payload to value
+    is needed to specify the type to convert to. Users can use type hints in their RBS files and such, but that has no
+    runtime check/effect.
+* The JSON encoding converter initializer will have a way to set parse and generate options.
+* The payload converter will have a way to let the default be created with JSON parse and generate options for the JSON
+  encoding converter.
+  * 💭 Why? Users wanting to simply set JSON options shouldn't have to go through the mess of recreating the default
+    list of encoding converters.
+
+## Activities
+
+A simple, heartbeating activity that calls a separate thing:
+
+```ruby
+class MyActivity < Temporalio::Activity
+  # Can customize name, e.g.
+  # activity_name 'custom-name'
+
+  # Can customize executor, e.g.
+  # activity_executor :fiber
+
+  # Can make an activity dynamic, e.g.
+  # activity_dynamic true
+
+  def initialize(db_client)
+    @db_client = db_client
+  end
+
+  def execute(arg)
+    loop do
+      db_result = @db_client.do_thing(arg)
+      if db_result
+        return db_result
+      end
+      sleep 1
+      Temporalio::Activity::Context.heartbeat
+    end
+  end
+end
+```
+
+* Activities are defined as classes.
+  * 💭 Why classes instead of methods/functions like every other SDK? There are tradeoffs to both approaches. As
+    classes, they are more easily referenced, a bit more easily typed in RBS, and metadata can be applied more easily at
+    the class level. This comes at a cost of not being able to easily share state across classes, but users can provide
+    something in the constructor multiple activities use. Open to considering activity methods instead.
```

Comment:
You might've already discussed  this internally, but a middle ground here might be supporting any object that responds to `#call` rather than `#execute`.

Using `#call` would potentially allow simple activities to be implemented as a `Proc` or `Lambda`, while still allowing people to use custom classes for more complex scenarios.

In the spirit of "there's more than one way to do it", this would give SDK users multiple options for defining activities, and they could choose whichever one is most appropriate for their circumstances:

```ruby
module Activities
  # Constant procs
  SomeProcActivity = ->(params) {
    # ...
  }

  # Proc factory methods
  def self.another_proc_activity(config)
    ->(params) {
      # can use config to customize proc
    }
  }

  # Maybe some kind of DSL?
  DSLActivity = Temporalio.activity(options) do |params|
    # ...
  end

  # Child classes for full control
  ChildClassActivity < Temporalio::Activity
    def initialize(config)
      # ...
    end

    def call(params)
      # ...
    end
  end
end
```

```ruby
# Activity registration using a hash, similar to TS API:
worker1 = Temporalio::Worker.new(
  # ...

  activities: {
    SomeProcActivity: Activities::SomeProcActivity,
    AnotherProcActivity: Activities.another_proc_activity(config),
    DSLActivity: Activities::DSLActivity,
    ChildClassActivity: Activities::ChildClassActivity,

    # Even inline activities, which is almost certainly not a good
    # idea for production code, but could be useful in some situations.
    InlineActivity: ->(params) {
      # ...
    }
  }

  # ...
)
```

And possibly a block-based API on the worker:

```ruby
worker.register_activity('MyActivity') do |params|
  # ...
end
```

As well as anything that responds to `#to_proc`:

```ruby
worker.register_activity('MyActivity', &:my_activity)
```

All of these could of course also be supported using the `#execute` method, but then there'd be a discrepancy between the SDK's `#execute` and the proc's `#call`.

---

Some similar APIs:
- the [Rack specification](https://github.com/rack/rack/blob/main/SPEC.rdoc), which makes heavy use of objects that respond to `#call`
- [`Logger#formatter`](https://ruby-doc.org/3.3.4/stdlibs/logger/Logger.html#attribute-i-formatter), which is documented as supported either a proc or an instance [`Logger::Formatter`](https://ruby-doc.org/3.3.4/stdlibs/logger/Logger/Formatter.html) (which responds to `#call`)
- AWS SDK uses [a `Callable` interface](https://github.com/search?q=repo%3Aaws%2Faws-sdk-ruby+%22%5Bcallable%5D%22&type=code) for various callback-like options in some services

Reply:
> Using #call would potentially allow simple activities to be implemented as a Proc or Lambda, while still allowing people to use custom classes for more complex scenarios.

A couple of important aspects we want for activities are 1) they are referenceable without calling (e.g. `Workflow.execute_activity(MyActivityClass, ...`), and 2) people can use existing things (e.g. mixins) and there is not much magic.

> In the spirit of "there's more than one way to do it", this would give SDK users multiple options for defining activities, 

We actually try to avoid this in SDKs where we can at declaration time. Sure some SDKs do have a couple of ways, but in general SDKs say "activities are defined as X" without many values for X and without more magic than we have to have. And remember there is the referenceability concern.

Reply:
> A couple of important aspects we want for activities are 1) they are easily referenceable by a caller without calling (e.g. Workflow.execute_activity(MyActivityClass, ..., w/ possible generic typing support w/ RBS/Sorbet), and 2) people can use existing things (e.g. mixins) and there is not much magic.

Got it. If I'm understanding them correctly, those all make sense to me as goals / requirements 👍

For (1), is that referring to the ability to call an activity with something like `MyActivity.execute` rather than `workflow.execute_activity(MyActivity)`? If so, then using a method other than `#call` definitely makes sense to me. IIUC, that'll help prevent users from trying to execute activity classes that haven't been wired up to Temporal.

> We actually try to avoid this in SDKs where we can at declaration time. Sure some SDKs do have a couple of ways, but in general SDKs say "activities are defined as X" without many values for X and without more magic than we have to have. And remember there is the referenceability concern.

I totally get this, and I'm _personally_ not a huge fan of how Ruby has adopted Perl's TIMTOWTDI philosophy. But I also think a lot of Rubyists definitely view that as one of Ruby's strengths, so just wanted to put in a plug for that way of thinking 😅

That being said, I don't think anyone will avoid using Temporal just because they can't implement activities as procs, so it's probably not worth the effort to implement (& support!) that level of flexibility.

Reply:
> is that referring to the ability to call an activity with something like `MyActivity.execute` rather than `workflow.execute_activity(MyActivity)`?

No, we definitely want the latter. We learned from other SDKs that it is best to not make it look like any local invocation.

> they can't implement activities as procs

Maybe we can have sugar for this, e.g. `class ProcActivity < Temporalio::Activity` and a class method on `Temporalio::Activity` that creates it. But it is the same type of sugar a user can write if we make sure that we support activities as classes.